typegen/src/utils.rs

```diff
@@ -76,110 +73,262 @@ pub fn ensure_unique_type_paths(types: &mut PortableRegistry) {
-    let ids_equal = |a: u32, b: u32| -> bool {
-        if a == b {
+
+    // Both types are recursive, and they look the same based on the above,
+    // so assume all is well, since we've already checked other things in prev recursion.
+    if seen_a && seen_b {
+        return true;
+    }
+
+    // Make note of whether these IDs (might) correspond to any specific generic.
+    let a_generic_idx = a_parent_params.index_for_type_id(a);
+    let b_generic_idx = b_parent_params.index_for_type_id(b);
+
+    // If both IDs map to same generic param, then we'll assume equal. If they don't
+    // then we need to keep checking other properties (eg Vec<bool> and Vec<u8> will have
+    // different type IDs but may be the same type if the bool+u8 line up to generic params).
+    if let (Some(a_idx), Some(b_idx)) = (a_generic_idx, b_generic_idx) {
+        if a_idx == b_idx {
             return true;
         }
-        let Some(a_param_names) = type_params_a.get(&a) else {
-            return false;
-        };
-        let Some(b_param_names) = type_params_b.get(&b) else {
-            return false;
-        };
-        // Check if there is any intersection, meaning that both IDs map to the same generic type param:
-        a_param_names.iter().any(|a_p| b_param_names.contains(a_p))
+    }
+
+    let a_ty = types.resolve(a).expect("type a should exist in registry");
+    let b_ty = types.resolve(b).expect("type b should exist in registry");
+
+    // Paths differ; types won't be equal then!
+    if a_ty.path.segments != b_ty.path.segments {
+        return false;
+    }
+
+    // Names of type params don't line up, so different then!
+    if a_ty
+        .type_params
+        .iter()
+        .zip(&b_ty.type_params)
+        .any(|(a, b)| a.name != b.name)
+    {
+        return false;
+    }
+
+    // We'll lazily extend our type params only if the shapes match.
+    let calc_params = || {
+        let a_params = a_parent_params.extend(&a_ty.type_params);
+        let b_params = b_parent_params.extend(&b_ty.type_params);
+        (a_params, b_params)
     };
 
-    let fields_equal = |a: &[Field<PortableForm>], b: &[Field<PortableForm>]| -> bool {
+    // Capture a few variables to avoid some repetition later when we recurse.
+    let mut types_equal_recurse =
+        |a: u32, a_params: &GenericsList, b: u32, b_params: &GenericsList| -> bool {
+            types_equal_inner(a, a_params, a_visited, b, b_params, b_visited, types)
+        };
+
+    // Check that all of the fields of some type are equal.
+    #[rustfmt::skip]
+    let mut fields_equal = |
+        a: &[Field<PortableForm>],
+        a_params: &GenericsList,
+        b: &[Field<PortableForm>],
+        b_params: &GenericsList
+    | -> bool {
         if a.len() != b.len() {
             return false;
         }
         a.iter().zip(b.iter()).all(|(a, b)| {
-            a.name == b.name && a.type_name == b.type_name && ids_equal(a.ty.id, b.ty.id)
+            a.name == b.name
```

Comment:
```suggestion
        let equal_name = a.name == b.name;

        if let (Some(a_type_name), Some(b_type_name)) = (&a.type_name, &b.type_name) {
            let type_names_present = a_params.index_for_type_id(a.ty.id).is_some()
                && b_params.index_for_type_id(b.ty.id).is_some();
            equal_name
                && a_type_name == b_type_name
                && !a_type_name.contains("::")
                && type_names_present
        } else {
            if a_params.index_for_type_id(a.ty.id) == b_params.index_for_type_id(b.ty.id)
                || a.type_name.as_ref().is_some_and(|x| x.contains("::"))
            {
                equal_name && types_equal_recurse(a.ty.id, a_params, b.ty.id, b_params)
            } else {
                dbg!(a, b);
                false
            }
        }
```

Reply:
this should not break any things, but the tests that were commented out should work

Reply:
I don't really follow the logic above offhand! Has our element conversation resolved this comment for you, or do you still think there's some improvement that we can make above?